src/Utilities/GenerateBatchVmSkus.cs

```diff
@@ -0,0 +1,163 @@
+                .Where(k => !configurationProperties.Contains(k, StringComparer.OrdinalIgnoreCase));
+
+            if (invalidArguments.Any())
+            {
+                throw new ArgumentException($"Invalid argument(s): {string.Join(", ", invalidArguments)}");
+            }
+
+            var configuration = new Configuration();
+            configurationSource.Bind(configuration);
+
+            return configuration;
+        }
+    }
+
+    internal class GenerateBatchVmSkus
+    {
+        static async Task Main(string[] args)
+        {
+            Configuration configuration = null;
+
+            try
+            {
+                configuration = Configuration.BuildConfiguration(args);
+            }
+            catch (ArgumentException ex)
+            {
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.WriteLine(ex.Message);
+                Console.ResetColor();
+                Environment.Exit(1);
+            }
+
+            Environment.Exit(await RunAsync(configuration));
+        }
+
+        static async Task<int> RunAsync(Configuration configuration)
+        {
+            var client = new ArmClient(new DefaultAzureCredential());
+            static double ConvertMiBToGiB(int value) => Math.Round(value / 1024.0, 2);
+            var subscription = client.GetSubscriptionResource(new ResourceIdentifier($"/subscriptions/{configuration.SubscriptionId}"));
+
+            var regionsForVm = new Dictionary<string, HashSet<string>>();
+            var sizeForVm = new Dictionary<string, VirtualMachineSize>();
+            var skuForVm = new Dictionary<string, ComputeResourceSku>();
+
+            foreach (var region in typeof(AzureLocation).GetTypeInfo().DeclaredProperties.Where(p => p.PropertyType == typeof(AzureLocation)).Select(p => p.Name).ToList())
+            {
+                try
+                {
+                    var vms = await subscription.GetBatchSupportedVirtualMachineSkusAsync(new AzureLocation(region)).Select(s => s.Name).ToListAsync();
+
+                    List<VirtualMachineSize>? sizes = null;
+                    List<ComputeResourceSku>? skus = null;
+                    foreach (var vm in vms)
+                    {
+                        if (regionsForVm.ContainsKey(vm))
+                        {
+                            regionsForVm[vm].Add(region);
+                        }
+                        else
+                        {
+                            regionsForVm[vm] = new HashSet<string>() { region };
+                        }
+
+                        if (!sizeForVm.ContainsKey(vm))
+                        {
+                            if (sizes is null)
+                            {
+                                sizes = await subscription.GetVirtualMachineSizesAsync(region).ToListAsync();
+                            }
+                            sizeForVm[vm] = sizes.FirstOrDefault(vmsize => vmsize.Name.Equals(vm));
```

Comment:
`SingleOrDefault` would be better here, so that it throws if more than 1 is found (which would tell us the data is dirty or we are not considering something)